### PR TITLE
[DPE-4261] Switch to split snap

### PR DESCRIPTION
**Pull request by dragomirp**
* Updating charm libs
* Adding issues noticed during wildcard db implementation in k8s
* Switching to the separate pgb snap

### CODE REVIEW

File: lib/charms/tempo_k8s/v1/charm_tracing.py [hunk 1]
```diff
@@ -232,60 +307,78 @@
...
+    _server_cert = getattr(charm_instance, server_cert_attr)
+    if callable(_server_cert):
+        server_cert = _server_cert()
+    else:
+        server_cert = _server_cert
 
     if server_cert is None:
         logger.warning(
-            f"{charm}.{server_cert_getter} returned None; sending traces over INSECURE connection."
+            f"{charm_type}.{server_cert_attr} is None; sending traces over INSECURE connection."
```
[the remark this answers is not in the document]
Comment (dragomirp) [reply lib/charms/tempo_k8s/v1/charm_tracing.py:351]: Attr name not a secret. It is also none.

File: src/charm.py [hunk 2]
```diff
@@ -219,9 +219,8 @@ def _on_install(self, _) -> None:
         # Try to disable pgbackrest service
         try:
             cache = snap.SnapCache()
-            selected_snap = cache["charmed-postgresql"]
+            selected_snap = cache[PGBOUNCER_SNAP_NAME]
             selected_snap.alias("psql")
```
Comment (dragomirp) [inline src/charm.py:223]: The first alias will be used if both snaps are installed in parallel

File: src/charm.py [hunk 3]
```diff
@@ -439,7 +438,9 @@ def _get_readonly_dbs(self, databases: Dict) -> Dict[str, str]:
         readonly_dbs = {}
         if self.backend.relation and "*" in databases:
             read_only_endpoints = self.backend.get_read_only_endpoints()
-            r_hosts = ",".join([r_host.split(":")[0] for r_host in read_only_endpoints])
+            sorted_rhosts = [r_host.split(":")[0] for r_host in read_only_endpoints]
+            sorted_rhosts.sort()
+            r_hosts = ",".join(sorted_rhosts)
```
Comment (dragomirp) [inline src/charm.py:443]: Sorting because read_only_endpoints is a set

File: src/charm.py [hunk 4]
```diff
@@ -644,7 +645,7 @@ def generate_relation_databases(self) -> Dict[str, Dict[str, Union[str, bool]]]:
             if "admin" in roles or "superuser" in roles or "createdb" in roles:
                 add_wildcard = True
         if add_wildcard:
-            databases["*"] = {"name": "*", "auth_dbname": database}
+            databases["*"] = {"name": "*", "auth_dbname": database, "legacy": False}
```
Comment (dragomirp) [inline src/charm.py:648]: Caused transient upgrade errors in k8s, adding for parity's sake.

File: src/relations/pgbouncer_provider.py [hunk 5]
```diff
@@ -126,7 +126,7 @@ def _on_database_requested(self, event: DatabaseRequestedEvent) -> None:
         dbs = self.charm.generate_relation_databases()
         dbs[str(event.relation.id)] = {"name": database, "legacy": False}
         roles = extra_user_roles.lower().split(",")
-        if "admin" in roles or "superuser" in roles:
+        if "admin" in roles or "superuser" in roles or "createdb" in roles:
```
Comment (dragomirp) [inline src/relations/pgbouncer_provider.py:129]: Forgotten create db here.

File: src/upgrade.py [hunk 6]
```diff
@@ -90,6 +90,7 @@ def _on_upgrade_granted(self, event: UpgradeGrantedEvent) -> None:
         if self.charm.unit.is_leader():
             self.charm.generate_relation_databases()
 
+        self.charm.create_instance_directories()
```
Comment (dragomirp) [inline src/upgrade.py:93]: Instance directories won't exist in the new snap' data dir

**Review by taurus-forever — APPROVED** [review]
Thank you! Great progress!

**Review by marceloneppel — COMMENTED** [review]
Code LGTM!

Maybe it is unrelated, but I saw the following blocked status consistently after upgrading from the edge channel revision to the revision built from this PR branch code:
```sh
pgbouncer/0*      blocked   idle            10.163.152.124  6432/tcp  PgBouncer service pgbouncer-pgbouncer-prometheus not running
```

Steps to reproduce:
```sh
juju deploy postgresql --channel 14/edge
juju deploy pgbouncer --channel 1/edge
juju deploy data-integrator --channel edge

juju config data-integrator database-name=test
juju relate postgresql pgbouncer
juju relate pgbouncer data-integrator

# wait for idle

juju run pgbouncer/leader pre-upgrade-check
juju refresh pgbouncer --path ./pgbouncer_ubuntu-22.04-amd64.charm
```

Also, I see that the `charmed-pgbouncer` snap is not installed after an upgrade:
```sh
ubuntu@juju-db1dc8-1:~$ sudo snap list
Name                Version        Rev    Tracking       Publisher        Notes
charmed-postgresql  14.11          113    14/stable      dataplatformbot  held
core20              20240416       2318   latest/stable  canonical✓       base
core22              20240408       1380   latest/stable  canonical✓       base
lxd                 5.0.3-d921d2e  28373  5.0/stable/…   canonical✓       -
snapd               2.63           21759  latest/stable  canonical✓       snapd
```

**Review by marceloneppel — APPROVED** [review]
LGTM! As discussed in MM, the upgrade will be handled in a different PR.

**Comment by dragomirp** [comment]
> LGTM! As discussed in MM, the upgrade will be handled in a different PR.

Issue to track upgrade bugs: https://github.com/canonical/pgbouncer-operator/issues/267

**Pull request merged**